lib/sync/pull-request.js

```diff
@@ -1,57 +1,91 @@
+function getNextPage(headers = {}) {
+  const nextUrl = ((headers.link || '').match(/<([^>]+)>;\s*rel="next"/) || [])[1];
+  if (!nextUrl) {
+    return;
+  }
+  const parsed = url.parse(nextUrl).query.split('&');
+  let nextPage;
+  parsed.forEach((query) => {
+    const [key, value] = query.split('=');
+    if (key === 'page') {
+      nextPage = Number(value);
+    }
+  });
+  return nextPage;
+}
+
 /**
  * @param {import('probot').GitHubAPI} github - The GitHub API Object
- * @param {any} repository - TBD
- * @param {any} cursor - TBD
+ * @param {RepositoryObject} repository - TBD
+ * @param {string|number} cursor - TBD
  * @param {any} perPage - TBD
  * @returns {any} TBD
  */
-exports.getPullRequests = async (github, repository, cursor, perPage) => {
+async function getPullRequests(github, repository, cursor, perPage) {
+  /** @type {number} */
+  let status;
+  /** @type {{link: string}} */
+  let headers;
+  /** @type {import('@octokit/rest').Octokit.PullsListResponse} */
   let edges;
+  if (!cursor) {
+    cursor = 1;
+  } else {
+    cursor = Number(cursor);
+  }
   const vars = {
     owner: repository.owner.login,
     repo: repository.name,
     per_page: perPage,
-    cursor,
+    page: cursor,
   };
 
   let asyncTags = [];
   await statsd.asyncTimer(
     async () => {
-      try {
-        ({ edges } = (await github.graphql(getPullRequestQuery, vars)).repository.pullRequests);
-        asyncTags.push('status:200');
-      } catch (error) {
-        logger.error('GraphQL Query Error for Sync Pull Request', error);
-        if (error.status) {
-          asyncTags.push(`status:${error.status}`);
-        } else if (error.httpError && error.httpError.status) {
-          asyncTags.push(`status:${error.httpError.status}`);
-        } else {
-          asyncTags.push('status:exception');
-        }
-        throw error;
-      }
+      ({ data: edges, status, headers } = (await github.pulls.list({
+        ...vars,
+        state: 'all',
+        sort: 'created',
+        direction: 'desc',
+      })));
+      asyncTags.push(`status:${status}`);
     },
-    'graphql.sync_pull_request',
+    'rest.sync_pull_request',
     1,
     asyncTags,
   )();
 
-  let pullRequests = await Promise.all(edges.map(async ({ node: pull }) => {
-    // Count of comments causes graphql to time out, fetch it from the regular API
-    const pr = await github.pulls.get({
-      owner: repository.owner.login,
-      repo: repository.name,
-      pull_number: pull.number,
-    });
-    pull.comments = { totalCount: pr.data.comments };
-    const { data } = transformPullRequest({ pull_request: pull, repository }, pull.author);
+  // Force us to go to a non-existant page if we're past the max number of pages
+  const nextPage = getNextPage(headers) || cursor + 1;
+  console.log(`CURSOR: ${cursor}, NEXTPAGE: ${nextPage}`);
+
+  edges.forEach((edge) => {
+    edge.cursor = nextPage;
+  });
+
+  let pullRequests = await Promise.all(edges.map(async (pull) => {
```

Comment:
Why do we paginate everything and then only take the first PR?

Reply:
this is one of those "mind blowing" things in there. the `transformPullRequest` builds this object that has an array with a single pull request in it, we return the data and that single PR and then it goes as an item in the `pullRequests` array. So we're using all the PRs but building them each individually in a really weird way.